File: src/go/141.linked-list-cycle.go
/*
 * @lc app=leetcode id=141 lang=golang
 *
 * [141] Linked List Cycle

 Given a linked list, determine if it has a cycle in it.

To represent a cycle in the given linked list, we use an integer pos which represents the position (0-indexed) in the linked list where tail connects to. If pos is -1, then there is no cycle in the linked list.



Example 1:

Input: head = [3,2,0,-4], pos = 1
Output: true
Explanation: There is a cycle in the linked list, where tail connects to the second node.

Example 2:

Input: head = [1,2], pos = 0
Output: true
Explanation: There is a cycle in the linked list, where tail connects to the first node.

Example 3:

Input: head = [1], pos = -1
Output: false
Explanation: There is no cycle in the linked list.

Follow up:

Can you solve it using O(1) (i.e. constant) memory?
*/

package main

import "fmt"

// ListNode 链表节点
type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	if head == nil {
		return false
	}
	slow, fast := head, head.Next
	for {
		if fast == nil {
			return false
		}
		if slow == fast {
			return true
		}
		slow = slow.Next
		fast = fast.Next
		if fast == nil {
			return false
		}
		fast = fast.Next
	}
}

func main() {
	node1 := &ListNode{Val: 1}
	node2 := &ListNode{Val: 2}
	node3 := &ListNode{Val: 3}
	node4 := &ListNode{Val: 4}
	node5 := &ListNode{Val: 5}
	node1.Next = node2
	node2.Next = node3
	node3.Next = node4
	node4.Next = node5
	node5.Next = node2
	b := hasCycle(node1)
	fmt.Printf("%v\n", b)
}

File: src/c/141.linked-list-cycle.c
struct ListNode {
  int val;
  struct ListNode *next;
};

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

bool hasCycle(struct ListNode *head) {
  if (head == NULL) {
    return false;
  }

  struct ListNode *slow = head;
  struct ListNode *fast = head->next;

  while (true) {
      if (fast == NULL) {
          return false;
      }
      if (slow == fast) {
          return true;
      }
      slow = slow->next;
      fast = fast->next;
      if (fast == NULL) {
          return false;
      }
      fast = fast->next;
  }
}
